File: backend/PruebaTecnicaChileautos.Core/Filters/CharacterFilter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PruebaTecnicaChileautos.Core.Filters
{
    public class CharacterFilter
    {
        public string? Name { get; set; }
        public string? Status { get; set; }
        public string? Species { get; set; }
        public string? Type { get; set; }
        public string? Gender { get; set; }


        public enum ValidStatus { 
            alive,
            dead,
            unknow
        }

        public string ToQueryString() {

            var queryParams = new List<string>();

            if (!string.IsNullOrWhiteSpace(Name))
                queryParams.Add($"name={Uri.EscapeDataString(Name)}");

            if (!string.IsNullOrWhiteSpace(Status) && Enum.TryParse<ValidStatus>(Status, out _))
                queryParams.Add($"status={Uri.EscapeDataString(Status)}");
            
            if (!string.IsNullOrWhiteSpace(Species))
                queryParams.Add($"species={Uri.EscapeDataString(Species)}");

            if (!string.IsNullOrWhiteSpace(Type))
                queryParams.Add($"type={Uri.EscapeDataString(Type)}");

            if (!string.IsNullOrWhiteSpace(Gender))
                queryParams.Add($"gender={Uri.EscapeDataString(Gender)}");

            return string.Join(",", queryParams);
        }
    }
}

File: frontend/rick-and-morty-app/src/app/core/services/models/character-filter.ts
export class CharacterFilter {
  name?: string;
  status?: string;
  species?: string;
  type?: string;
  gender?: string;

  toQueryString(): string {
    const params: string[] = [];

    if (!!this.name) params.push(`name=${encodeURIComponent(this.name)}`);
    if (!!this.status) params.push(`status=${encodeURIComponent(this.status)}`);
    if (!!this.species) params.push(`species=${encodeURIComponent(this.species)}`);
    if (!!this.type) params.push(`type=${encodeURIComponent(this.type)}`);
    if (!!this.gender) params.push(`gender=${encodeURIComponent(this.gender)}`);

    return params.join('&');
  }
}